### PR TITLE
Problem: memiavl background snapshot writing don't limit concurrency level

### PR DESCRIPTION
**Pull request by jacksoom**
👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: memiavl/multitree.go [hunk 1]
```diff
@@ -364,6 +372,8 @@ func (t *MultiTree) WriteSnapshot(dir string) error {
 
 	// write the snapshots in parallel
 	g, _ := errgroup.WithContext(context.Background())
+	g.SetLimit(t.snapshotWriterLimit)
```
Comment (yihuang) [inline memiavl/multitree.go:375]: will it be cleaner if we add this as parameter in `WriteSnapshot` method, so we don't need to add a field in `MultiTree`.

File: memiavl/db.go [hunk 2]
```diff
@@ -898,8 +902,8 @@ func walPath(root string) string {
 //
 // current -> snapshot-0
 // ```
-func initEmptyDB(dir string, initialVersion uint32) error {
-	tmp := NewEmptyMultiTree(initialVersion, 0)
+func initEmptyDB(dir string, initialVersion uint32, snapshotWriterLimit int) error {
```
Comment (yihuang) [inline memiavl/db.go:905]: we can use a constant in this method to keep it simpler.

Comment (jacksoom) [reply memiavl/db.go]: > we can use a constant `1` in this method to keep it simpler.

fix

File: memiavl/multitree.go (hunk 1 above)
Comment (jacksoom) [reply memiavl/multitree.go]: > will it be cleaner if we add this as parameter of `WriteSnapshot` method, so we don't need to add a field in `MultiTree`.

Fixed

**Review by mmsqe — COMMENTED** [review]
maybe could limit with CPU-bound like `nProcs := runtime.GOMAXPROCS(0)`

File: memiavl/multitree.go [hunk 3]
```diff
@@ -88,7 +88,7 @@ func LoadMultiTree(dir string, zeroCopy bool, cacheSize int) (*MultiTree, error)
 		treeMap[name] = NewFromSnapshot(snapshot, zeroCopy, cacheSize)
 	}
 
-	sort.Strings(treeNames)
+	slices.Sort(treeNames)
```
Comment (jacksoom) [inline memiavl/multitree.go:91]: golang advice: consider using the newer slices.Sort function, which runs faster.

**Comment by jacksoom** [comment]
> maybe could limit with CPU-bound like `nProcs := runtime.GOMAXPROCS(0)`

Just to limit excessive resource usage for snapshot writing func. not to limit the main process

**Comment by jacksoom** [comment]
> maybe could limit with CPU-bound like `nProcs := runtime.GOMAXPROCS(0)`



> > maybe could limit with CPU-bound like `nProcs := runtime.GOMAXPROCS(0)`
> 
> Just to limit excessive resource usage for snapshot writing func (does not affect other goroutines). not to limit the main process

**Comment by jacksoom** [comment]
> maybe could limit with CPU-bound like `nProcs := runtime.GOMAXPROCS(0)`

只是防止影响其他goroutine

File: memiavl/db.go [hunk 4]
```diff
@@ -968,7 +974,7 @@ func atomicRemoveDir(path string) error {
 }
 
 // createDBIfNotExist detects if db does not exist and try to initialize an empty one.
-func createDBIfNotExist(dir string, initialVersion uint32) error {
+func createDBIfNotExist(dir string, initialVersion uint32, snapshotWorkerNum int) error {
```
Comment (yihuang) [inline memiavl/db.go:977]: The new parameter is not used

File: memiavl/db.go [hunk 5]
```diff
@@ -97,6 +99,8 @@ type Options struct {
 	// truncate the versions after the `TargetVersion`, the `TargetVersion` becomes the latest version.
 	// it do nothing if the target version is `0`.
 	LoadForOverwriting bool
+
+	SnapshotWriterLimit int
```
Comment (yihuang) [inline memiavl/db.go:103]: In fillDefaults, should we reset it to the default value if it's zero or negative?

**Review by yihuang — COMMENTED** [review]
LGTM, Minor suggestions.

**Review by mmsqe — APPROVED** [review]
LGTM, would you mind update gomod2nix.toml and sign the commits?

**Comment by jacksoom** [comment]
> LGTM, would you mind update gomod2nix.toml and sign the commits?

ok

**Comment by jacksoom** [comment]
> LGTM, would you mind update gomod2nix.toml and sign the commits?

done

**Comment by yihuang** [comment]
@jacksoom you need to sign all commits, you can squash the commits instead.

File: memiavl/db_test.go [hunk 6]
```diff
@@ -16,8 +16,9 @@ import (
 
 func TestRewriteSnapshot(t *testing.T) {
 	db, err := Load(t.TempDir(), Options{
-		CreateIfMissing: true,
-		InitialStores:   []string{"test"},
+		CreateIfMissing:     true,
+		InitialStores:       []string{"test"},
+		SnapshotWriterLimit: DefaultSnapshotWriterLimit,
```
Comment (yihuang) [inline memiavl/db_test.go:21]: we can remove the default values in the test, because it'll be filled by default.

**Review by yihuang — APPROVED** [review]
LGTM, minor suggestions to void adding default option to the test.

**Pull request closed without merging**